sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java

```diff
@@ -446,4 +504,164 @@ public void close() throws IOException {
+      abstract Write<T> build();
+    }
+
+    public Write<T> withUri(String uri) {
+      checkNotNull(uri);
+      return toBuilder().setUri(uri).build();
+    }
+
+    public Write<T> withDatabase(String database) {
+      checkNotNull(database);
+      return toBuilder().setDatabase(database).build();
+    }
+
+    public Write<T> withBucket(String bucket) {
+      checkNotNull(bucket);
+      return toBuilder().setBucket(bucket).build();
+    }
+
+    public Write<T> withFilename(String filename) {
+      checkNotNull(filename);
+      return toBuilder().setFilename(filename).build();
+    }
+
+    public Write<T> withChunkSize(Long chunkSize) {
+      checkNotNull(chunkSize);
+      checkArgument(chunkSize > 1, "Chunk Size must be greater than 1", chunkSize);
+      return toBuilder().setChunkSize(chunkSize).build();
+    }
+
+    public void validate(T input) {
+      checkNotNull(filename(), "filename");
+      checkNotNull(writeFn(), "writeFn");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder.addIfNotNull(DisplayData.item("uri", uri()));
+      builder.addIfNotNull(DisplayData.item("database", database()));
+      builder.addIfNotNull(DisplayData.item("bucket", bucket()));
+      builder.addIfNotNull(DisplayData.item("filename", filename()));
+    }
+
+    @Override
+    public PDone apply(PCollection<T> input) {
+      input.apply(ParDo.of(new GridFsWriteFn<T>(this)));
+      return PDone.in(input.getPipeline());
+    }
+  }
+  private static class GridFsWriteFn<T> extends DoFn<T, Void> {
+
+    private final Write<T> spec;
+
+    private Mongo mongo;
+    private GridFS gridfs;
+
+    private GridFSInputFile gridFsFile;
+    private OutputStream outputStream;
+
+    public GridFsWriteFn(Write<T> spec) {
+      this.spec = spec;
+    }
+
+    private Mongo setupMongo() {
+      return spec.uri() == null ? new Mongo() : new Mongo(new MongoURI(spec.uri()));
+    }
+
+    private GridFS setupGridFS(Mongo mongo) {
+      DB db = spec.database() == null ? mongo.getDB("gridfs") : mongo.getDB(spec.database());
+      return spec.bucket() == null ? new GridFS(db) : new GridFS(db, spec.bucket());
+    }
+
+
+    @Setup
+    public void setup() throws Exception {
+      mongo = setupMongo();
+      gridfs = setupGridFS(mongo);
+    }
+
+    @StartBundle
+    public void startBundle(Context context) {
+      gridFsFile = gridfs.createFile(spec.filename());
+      if (spec.chunkSize() != null) {
+        gridFsFile.setChunkSize(spec.chunkSize());
+      }
+      outputStream = gridFsFile.getOutputStream();
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext context) throws Exception {
+      T record = context.element();
+      spec.writeFn().write(record, outputStream);
+    }
+
+    @FinishBundle
+    public void finishBundle(Context context) throws Exception {
+      if (gridFsFile != null) {
```

Comment:
That should be fine.  Whenever the "chunkSize" is hit (which is configurable), the chunk is persisted to the database and file updated and such.